cpp/dolfin/fem/DirichletBC.cpp

```diff
@@ -495,125 +234,84 @@ void DirichletBC::dof_values(
-//             continue;
-
-//           // Skip already checked dofs
-//           if (already_visited.in_range(cell_dofs[i])
-//               and !already_visited.insert(cell_dofs[i]))
-//           {
-//             continue;
-//           }
-
-//           bc_dofs.push_back(cell_dofs[i]);
-//           bc_dofs_g.push_back(cell_dofs_g[i]);
-//         }
-//       }
-//     }
-//   }
-
-//   // FIXME: Send to other (neigbouring) processes, maybe just for shared
-//   // dofs?
-
-//   return std::set<PetscInt>(bc_dofs.begin(), bc_dofs.end());
-// }
+Eigen::Array<PetscInt, Eigen::Dynamic, 1> fem::locate_dofs_topological(
+    const function::FunctionSpace& V, const int entity_dim,
+    const Eigen::Ref<const Eigen::Array<PetscInt, Eigen::Dynamic, 1>>& entities)
+{
+
+  const DofMap& dofmap = *V.dofmap();
+  dolfin::mesh::Mesh mesh = *V.mesh();
+
+  const int tdim = mesh.topology().dim();
+
+  // Initialise entity-cell connectivity
+  mesh.create_entities(tdim);
+  mesh.create_connectivity(entity_dim, tdim);
+
+  // Prepare an element-local dof layout for dofs on entities of the entity_dim
+  const int num_entities
+      = mesh::cell_num_entities(mesh.cell_type(), entity_dim);
+  std::vector<Eigen::Array<int, Eigen::Dynamic, 1>> entity_dofs;
+  for (int i = 0; i < num_entities; ++i)
+  {
+    entity_dofs.push_back(
+        dofmap.element_dof_layout->entity_closure_dofs(entity_dim, i));
+  }
+
+  const std::size_t num_entity_closure_dofs
+      = dofmap.element_dof_layout->num_entity_closure_dofs(entity_dim);
+
+  std::vector<PetscInt> dofs;
+
+  for (Eigen::Index i = 0; i < entities.rows(); ++i)
+  {
+    // Create entity and attached cell
+    const mesh::MeshEntity entity(mesh, entity_dim, entities[i]);
+    const std::size_t cell_index = entity.entities(tdim)[0];
+    const mesh::MeshEntity cell(mesh, tdim, cell_index);
+
+    // Get cell dofmap
+    auto cell_dofs = dofmap.cell_dofs(cell_index);
+
+    // Loop over entity dofs
+    const size_t entity_local_index = cell.index(entity);
+    for (std::size_t j = 0; j < num_entity_closure_dofs; j++)
+    {
+      const std::size_t index = entity_dofs[entity_local_index][j];
+      const PetscInt dof_index = cell_dofs[index];
+      dofs.push_back(dof_index);
+    }
+  }
+
+  return Eigen::Map<Eigen::Array<PetscInt, Eigen::Dynamic, 1>>(dofs.data(),
+                                                               dofs.size());
+}
 //-----------------------------------------------------------------------------
+Eigen::Array<PetscInt, Eigen::Dynamic, 1>
+fem::locate_dofs_geometrical(const function::FunctionSpace& V,
+                             marking_function marker)
+{
+
+  const Eigen::Array<double, 3, Eigen::Dynamic, Eigen::RowMajor> dof_coordinates
+      = V.tabulate_dof_coordinates().transpose();
+  const Eigen::Array<bool, Eigen::Dynamic, 1> marked_dofs
+      = marker(dof_coordinates);
+
+  std::vector<PetscInt> dofs;
+
+  for (PetscInt i = 0; i < marked_dofs.size(); ++i)
```

Comment:
What's this got to do with Petsc?

Reply:
What type do you suggest? `std::int32_t` since this are just local dof indices?

Reply:
It should be whatever type ` marked_dofs.size()` is.